### ISSUE TITLE
Summary generation for tensorboard is very slow

### ISSUE THREAD

**Issue by ipcoder** [pytorch/pytorch#35188]
I have noticed that adding even a single scalar to log with SummaryWriter slows by~10 the learning rate:
```
tb.add_scalar('loss', loss, global_step=self.train_step)
```
After some experiments and examination of the implementation I have found that **even without writing the data**, only  the __creation__ of the Summary item takes ~.5 sec.
```
t0 = time()
s = Summary(value=[Summary.Value(tag='loss', simple_value=float(loss))])
print(time()-t0)
```
This is of course a very rough measurement but it is consistent with  ~10% drop in GPU load.
## Expected behavior

I would expect much less time for a single scalar summary item _creation_, few ms perhaps or even less.

## Environment
```
 - PyTorch Version (e.g., 1.4.1):
 - Kubuntu Linux:
 - installed PyTorch from `conda`:
 - Build command you used (if compiling from source):
 - Python version: 3.6.10
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: GeForce RTX 2080 Ti
 - Any other relevant information: 
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 440.64.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.4

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] numpydoc==0.9.2
[pip] torch==1.4.0
[pip] torchvision==0.5.0
[conda] _pytorch_select           0.1                       cpu_0  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.0                      166  
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.0.15           py36ha843d7b_0  
[conda] mkl_random                1.1.0            py36hd6b4f25_0  
[conda] pytorch                   1.4.0           py3.6_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.5.0                py36_cu101    pytorch
```

**Comment by ipcoder** [comment]
I have figured out that the reason for the slow down is the access to the loss tensor on the CUDA - not related to the tensorboard interface.

**Comment by danpovey** [comment]
Guys, I think perhaps this issue could be reopened.  After writing a more-than-usual  amount of things to a tensorboard with a log-file on a remote file system and finding it super super slow, I wrote a test program to debug it and found something interesting.  Every single time it writes an item to the tensorboard, it re-opens the file, seeks to a certain position, writes the protobuf and then closes the file.  So no wonder it's slow when the file is on a remote file system.  Absolutely zero buffering and doesn't even keep an open filehandle.

The following strace log is with torch version 1.8.1+cu101 and tensorboard version 2.9.0,  but I have verified that the same happens with torch 2.1.0+cu121 and tensorboard 2.14.0; and also with 2.2.2+cu121 and tensorboard 2.17.0.
The reason I do "sleep" at the end is because if I don't do that, nothing gets written at all.  Yah, tensorboard is janky.
```
import torch
import time
from torch.utils.tensorboard import SummaryWriter


tb_writer = SummaryWriter(log_dir=f"file://tensorboard",
                          max_queue=100)
for i in range(100):
    tb_writer.add_scalar("train/epoch", i, i)

time.sleep(10)
```
I then did:  `strace -f python3 test.py &> strace`
```
[pid 59105] openat(AT_FDCWD, "/tmp/tensorboard/events.out.tfevents.1741330231.de-74279-k2-train-7-0905180047-6d6678bc6f-8cwvw.57601.0", O_WRONLY|O_CREAT|O_APPEND|O_CLOEXEC, 0666) = 3
[pid 59105] fstat(3, {st_mode=S_IFREG|0644, st_size=4693, ...}) = 0
[pid 59105] lseek(3, 0, SEEK_END)       = 4693
[pid 59105] ioctl(3, TCGETS, 0x7f914b0bef20) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 59105] lseek(3, 0, SEEK_CUR)       = 4693
[pid 59105] write(3, "!\0\0\0\0\0\0\0\1{\354\366\t\3535\321\315\245\362\331A\20_*\24\n\22\n\vtra"..., 49) = 49
[pid 59105] close(3)                    = 0
[pid 59105] openat(AT_FDCWD, "/tmp/tensorboard/events.out.tfevents.1741330231.de-74279-k2-train-7-0905180047-6d6678bc6f-8cwvw.57601.0", O_WRONLY|O_CREAT|O_APPEND|O_CLOEXEC, 0666) = 3
[pid 59105] fstat(3, {st_mode=S_IFREG|0644, st_size=4742, ...}) = 0
[pid 59105] lseek(3, 0, SEEK_END)       = 4742
[pid 59105] ioctl(3, TCGETS, 0x7f914b0bef20) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 59105] lseek(3, 0, SEEK_CUR)       = 4742
[pid 59105] write(3, "!\0\0\0\0\0\0\0\1{\354\366\t\2467\321\315\245\362\331A\20`*\24\n\22\n\vtra"..., 49) = 49
[pid 59105] close(3)                    = 0
[pid 59105] openat(AT_FDCWD, "/tmp/tensorboard/events.out.tfevents.1741330231.de-74279-k2-train-7-0905180047-6d6678bc6f-8cwvw.57601.0", O_WRONLY|O_CREAT|O_APPEND|O_CLOEXEC, 0666) = 3
[pid 59105] fstat(3, {st_mode=S_IFREG|0644, st_size=4791, ...}) = 0
[pid 59105] lseek(3, 0, SEEK_END)       = 4791
[pid 59105] ioctl(3, TCGETS, 0x7f914b0bef20) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 59105] lseek(3, 0, SEEK_CUR)       = 4791
[pid 59105] write(3, "!\0\0\0\0\0\0\0\1{\354\366\t\0339\321\315\245\362\331A\20a*\24\n\22\n\vtra"..., 49) = 49
[pid 59105] close(3)                    = 0
[pid 59105] openat(AT_FDCWD, "/tmp/tensorboard/events.out.tfevents.1741330231.de-74279-k2-train-7-0905180047-6d6678bc6f-8cwvw.57601.0", O_WRONLY|O_CREAT|O_APPEND|O_CLOEXEC, 0666) = 3
[pid 59105] fstat(3, {st_mode=S_IFREG|0644, st_size=4840, ...}) = 0
[pid 59105] lseek(3, 0, SEEK_END)       = 4840
[pid 59105] ioctl(3, TCGETS, 0x7f914b0bef20) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 59105] lseek(3, 0, SEEK_CUR)       = 4840
[pid 59105] write(3, "!\0\0\0\0\0\0\0\1{\354\366\tq:\321\315\245\362\331A\20b*\24\n\22\n\vtra"..., 49) = 49
[pid 59105] close(3)                    = 0
[pid 59105] openat(AT_FDCWD, "/tmp/tensorboard/events.out.tfevents.1741330231.de-74279-k2-train-7-0905180047-6d6678bc6f-8cwvw.57601.0", O_WRONLY|O_CREAT|O_APPEND|O_CLOEXEC, 0666) = 3
[pid 59105] fstat(3, {st_mode=S_IFREG|0644, st_size=4889, ...}) = 0
[pid 59105] lseek(3, 0, SEEK_END)       = 4889
[pid 59105] ioctl(3, TCGETS, 0x7f914b0bef20) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 59105] lseek(3, 0, SEEK_CUR)       = 4889
[pid 59105] write(3, "!\0\0\0\0\0\0\0\1{\354\366\t\354<\321\315\245\362\331A\20c*\24\n\22\n\vtra"..., 49) = 49
[pid 59105] close(3)                    = 0
```
The time where this was slow was with a StarFS file system, mounted from a remote machine, but the above strace is from when the file system was on /tmp (of type "overlay" as reported by "df"), just to demonstrate that it does this opening-and-closing thing regardless of what type of file system it is.
The ridiculous thing is that these tensorboard files are only accessed by a single process anyway, as tensorboard creates them and accesses them with a single dedicated thread, so there's no obvious reason why it would need to close and re-open each time.  And you can't justify this by saying it needs to make sure the messages get flushed to disk if the process exits, because as you can see from my need to "sleep" at exit, it just discards any queued messages when the program exits anyway.
Unfortunately it would be hard to fix this from the Python side without a refactor, since the code that chooses to use Tensorflow's GFile instead of a normal python file is inside tensorboard.
